espressostreamer/espresso_streamer.go

```diff
@@ -96,33 +99,62 @@ func NewEspressoStreamer(namespace uint64, hotshotUrls []string,
 		l1Reader:                      l1Reader,
 		espressoTEEVerifierABI:        espressoTEEVerifierAbi,
 		espressoTEEVerifierAddr:       espressoTEEVerifierAddress,
+		// Initally, we set the confirmed message pos to max uint64
+		// because we dont know the confirmed message pos yet
+		confirmedMessagePos: abi.MaxUint256.Uint64(),
 	}
 }
 
-/*
-Pop the first message from the queue
-It will return nil if the queue is empty
-*/
-func (s *EspressoStreamer) PopMessageWithMetadataAndPos() *MessageWithMetadataAndPos {
-	if len(s.messageWithMetadataAndPos) == 0 {
-		return nil
+func (s *EspressoStreamer) Refresh(ctx context.Context, fetchLastMessageAndEspressoBlock func() (uint64, uint64, error)) (bool, error) {
+	lastMessagePos, lastEspressoBlock, err := fetchLastMessageAndEspressoBlock()
+	if err != nil {
+		return false, err
+	}
+
+	log.Info("Refreshing espresso streamer", "lastMessagePos", lastMessagePos, "lastEspressoBlock", lastEspressoBlock, "confirmedMessagePos", s.confirmedMessagePos)
+	// If the last message pos only increase by 1, we dont need to refresh
+	// because this is expected behavior
+	if lastMessagePos == s.confirmedMessagePos+1 {
+		s.confirmedMessagePos = lastMessagePos
+		s.currentMessagePos = lastMessagePos + 1
+		return false, nil
 	}
-	messageWithMetadataAndPos := s.messageWithMetadataAndPos[0]
-	s.messageWithMetadataAndPos = s.messageWithMetadataAndPos[1:]
-	return messageWithMetadataAndPos
+
+	// If last message pos didnt increase by and if not equal to the confirmed message pos
+	// something went wrong
+	// we reset the values to the values returned from the source of truth.
+	if lastMessagePos != s.confirmedMessagePos {
+		s.confirmedMessagePos = lastMessagePos
+		s.currentMessagePos = lastMessagePos + 1
+		s.nextHotshotBlockNum = lastEspressoBlock
+		s.messageWithMetadataAndPos = []*MessageWithMetadataAndPos{}
+		return true, nil
+	}
+
+	return false, nil
 }
 
-/*
-Peek the first message from the queue
-It will return nil if the queue is empty
-This will be used to check if the the EspressoStreamer has recieved
-the next expected message in sequence
-*/
-func (s *EspressoStreamer) PeekMessageWithMetadataAndPos() *MessageWithMetadataAndPos {
-	if len(s.messageWithMetadataAndPos) == 0 {
-		return nil
+func (s *EspressoStreamer) Next() (*MessageWithMetadataAndPos, error) {
+
+	// Find if the current message pos is in the queue
+	// delete any messages before the current message pos
+	for i := 0; i < len(s.messageWithMetadataAndPos); i++ {
+		if s.messageWithMetadataAndPos[i].Pos < s.currentMessagePos {
+			s.messageWithMetadataAndPos = s.messageWithMetadataAndPos[i+1:]
+			continue
+		}
+		break
+	}
+
+	if len(s.messageWithMetadataAndPos) > 0 && s.messageWithMetadataAndPos[0].Pos == (s.currentMessagePos) {
+		// Remove the first element
+		msg := s.messageWithMetadataAndPos[0]
+		s.messageWithMetadataAndPos = s.messageWithMetadataAndPos[1:]
+		s.currentMessagePos++
+		return msg, nil
```

Comment:
I think in this process, we can just find the next position and we don't have to sort the slice anymore.

I would suggest that we can interate the memory queue until we find the next position and remove those outdated iterated ones, which should result in a better performance

Reply:
We still need sorting because otherwise, think about this scenario if we dont sort. 
The array can be [4,3, 2] and the message position we are looking for is 3. Then we will end up deleting 4 as well, if we dont sort which is not the right behaviour.

Reply:
If we are looking for 3 when the queue is [4, 3, 2], then we can keep visiting the queue until we get 3. And if we encounter less number than 3, we can delete it.

the queue will be like:
- [4, 2] // consumes 3 and keeps 4 left
- [2] // consuemes 4
- [] // looking for 5 and removing 2 because 2 < 5

Reply:
okay makes sense, I think you are talking about the same thing [here](https://github.com/EspressoSystems/nitro-espresso-integration/pull/479#discussion_r1960853711) as well. Feel free to implement this :)